packages/shared-action-utils/src/publish.ts

```diff
@@ -0,0 +1,126 @@
+
+const globAsync = (pattern: string, options: IOptions): Promise<string[]> => {
+    return new Promise<string[]>((resolve, reject) => {
+        glob(pattern, options, (error, matches) => {
+            if (error) {
+                reject(error);
+            } else {
+                resolve(matches);
+            }
+        });
+    });
+};
+
+type PackageJson = {
+    name: string;
+    version: string;
+};
+
+const loadProjectPackageJson = async (packageJsonPath: string): Promise<PackageJson> => {
+    if (!existsSync(packageJsonPath)) {
+        throw new Error(`Could not locate package.json in ${packageJsonPath}`);
+    }
+
+    const contents = await readFile(packageJsonPath);
+
+    return JSON.parse(contents.toString());
+};
+
+type GitHubAction = {
+    name: string;
+    version: string;
+    tag: string;
+    releaseBranch: string;
+    directoryPath: string;
+};
+
+async function gatherActionsToPotentiallyPublish(): Promise<Array<GitHubAction>> {
+    const actions = await globAsync("**/action.yml", { ignore: ["packages/**", "node_modules/**", "**/__tests/**"] });
+
+    const actionsToPublish: Array<GitHubAction> = [];
+
+    for (const actionPath of actions) {
+        // Get the folder of the action which doubles as the key
+        // e.g. for {reporoot}/extract-package-details/action.yml
+        // the action key is extrac-package-details
+        let actionPathFolders = actionPath.split("/");
+        actionPathFolders = actionPathFolders.slice(0, actionPathFolders.length - 1);
+        const actionFolder = actionPathFolders[actionPathFolders.length - 1];
+        const packageJsonPathForAction = `./packages/${actionFolder}/package.json`;
+
+        const packageJsonForAction = await loadProjectPackageJson(packageJsonPathForAction);
+
+        actionsToPublish.push({
+            name: packageJsonForAction.name,
+            version: packageJsonForAction.version,
+            tag: `${packageJsonForAction.name}@${packageJsonForAction.version}`,
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            releaseBranch: `v${packageJsonForAction.version.split(".")[0]!}`,
+            directoryPath: path.join(...actionPathFolders),
+        });
+    }
+
+    return actionsToPublish;
+}
+
+async function publishActionIfRequired(action: GitHubAction): Promise<void> {
+    const { exitCode, stderr } = await getExecOutput(`git`, ["ls-remote", "--exit-code", "origin", "--tags", `refs/tags/${action.tag}`], {
+        ignoreReturnCode: true,
+    });
+    if (exitCode === 0) {
+        console.log(`Action is not being published because version ${action.tag} is already published`);
+        return;
+    }
+    if (exitCode !== 2) {
+        throw new Error(`git ls-remote exited with ${exitCode}:\n${stderr}`);
+    }
+
+    console.log(`Publishing action '${action.name}' version '${action.version}' to distribution branch '${action.releaseBranch}'`);
+
+    await exec("git", ["checkout", "--detach"]);
```

Comment:
This publish script runs within the context of an existing branch (typically `main`), and doing the `git checkout --detach` means we still have all the commits from the source branch we just aren't adding to them (as far as I understand anyway) as we are in a detached state. We then commit the new distributable js file and push it up connected to the tag.

An implication of this is that publishing any version of an action would effectively bring all commits from main at the time into the distribution branch (e.g. v1). This means that the source on the distribution branch is not really what the js was built from. This feels like it could be a bad thing.

Reply:
In addition it feels like this kind of breaks the "must be able to get from a version number through to the code for that version" rule as we don't tag the commit that the code was built from we tag the commit that contains the built code. I guess we can go back to the previous "non-release" commit that triggered the build and that would be the source? Still non-obvious.